core/src/main/scala/com/github/johnynek/paiges/Doc.scala

```diff
@@ -124,13 +122,42 @@ sealed abstract class Doc extends Serializable {
       curr * 1500450271 + c.toInt
     @tailrec def shash(n: Int, s: String, i: Int): Int =
       if (i < s.length) shash(hash(n, s.charAt(i)), s, i + 1) else n
-    Tok.fromDoc(this).foldLeft(0xdead60d5) {
+    Tok.fromDoc(inner).foldLeft(0xdead60d5) {
       case (n, Tok.Line(i)) => hash(n, '\n') + (1500450271 * i)
       case (n, Tok.Text(s)) => shash(n, s, 0)
     }
   }
 
-  override def toString: String = "Doc(...)"
+  override def toString: String = {
+    var remaining = 40
+    val bldr = new java.lang.StringBuilder
+    val r = renderStream(80)
+      .filter(_.nonEmpty)
+      .takeWhile { str =>
+        if (remaining == 0) {
+          // signal that we have more to write
+          bldr.append("...")
+          remaining = -1
+          false
+        }
+        else {
+          val tooMuch = remaining < str.length
+          val substring = if (tooMuch) str.substring(0, remaining) else str
+          remaining -= substring.length
+          bldr.append(substring)
+          if (tooMuch) { bldr.append("...") }
+          true
+        }
+      }
+      .foreach(_ => ()) // side effect already happened
+
+    s"Doc(${bldr.toString})##$hashCode"
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case d: Doc => (this eq d) || (inner == d.inner) || (compare(d) == 0)
```

Comment:
Are you sure it's worth doing the structural test before the comparison? In the case of most differences, I don't expect it to be significantly faster than `compare`, and when they are equal it will be significantly slower.